### PR TITLE
refactor(ir): split the relational operations

### PR DESCRIPTION
**Pull request by kszucs**
### Rationale & History

In the last couple of years we have been constantly refactoring the internals to make it easier to work with. 
Although we have made great progress, the current codebase is still hard to maintain and extend. One example of that complexity is the try to remove the `Projector` class in #7430. I had to realize that we are unable to improve the internals in smaller incremental steps, we need to make a big leap forward to make the codebase maintainable in the long run.

One of the hotspots of problems is the `analysis.py` module which tries to bridge the gap between the user-facing API and the internal representation. Part of its complexity is caused by loose integrity checks in the internal representation, allowing various ways to represent the same operation. This makes it hard to inspect, reason about and optimize the relational operations. In addition to that, it makes much harder to implement the backends since more branching is required to cover all the variations. 

We have always been aware of these problems, and actually we had several attempts to solve them the same way this PR does. However, we never managed to actually split the relational operations, we always hit roadblocks to maintain compatibility with the current test suite. Actually we were unable to even understand those issues because of the complexity of the codebase and number of indirections between the API, analysis functions and the internal representation. 

But(!) finally we managed to prototype a new IR in #7580 along with implementations for the majority of the backends, including `various SQL backends` and `pandas`. After successfully validating the viability of the new IR, we split the PR into smaller pieces which can be individually reviewed. This PR is the first step of that process, it introduces the new IR and the new API. The next steps will be to implement the remaining backends on top of the new IR.

### What does the PR do:

- Split the `ops.Selection` and `ops.Aggregration` nodes into proper relational algebra operations.
- Almost entirely remove `analysis.py` with the technical debt accumulated over the years.
- More flexible window frame binding: if an unbound analytical function is used with a window containing references to a relation then `.over()` is now able to bind the window frame to the relation.
- Introduce a new API-level technique to dereference columns to the target relation(s).
- Revamp the subquery handling to be more robust and to support more use cases with strict validation, now we have `ScalarSubquery`, `ExistsSubquery`, and `InSubquery` nodes which can only be used in the appropriate context.
- Use way stricter integrity checks for all the relational operations, most of the time enforcing that all the value inputs of the node must originate from the parent relation the node depends on.
- Introduce a new `JoinChain` operations to represent multiple joins in a single operation followed by a projection attached to the same relation. This enabled to solve several outstanding issues with the join handling (including the notorious chain join issue).
- Use straightforward rewrite rules collected in `rewrites.py` to reinterpret user input so that the new operations can be constructed, even with the strict integrity checks.
- Provide a set of simplification rules to reorder and squash the relational operations into a more compact form.
- Use mappings to represent projections, eliminating the need of internally storing `ops.Alias` nodes. In addition to that table nodes in projections are not allowed anymore, the columns are expanded to the same mapping making the semantics clear.
- Uniform handling of the various kinds of inputs for all the API methods using a generic `bind()` function.

#### Advantages:
- The operations are much simpler with clear semantics.
- The operations are easier to reason about and to optimize.
- The backends can easily lower the internal representation to a backend-specific form before compilation/execution, so the lowered form can be easily inspected, debugged, and optimized.
- The API is much closer to the users' mental model, thanks to the dereferencing technique.
- The backend implementation can be greatly simplified due to the simpler internal representation and strict integrity checks. As an example the pandas backend can be slimmed down by 4k lines of code while being more robust and easier to maintain.

#### Disadvantages:
- The backends must be rewritten to support the new internal representation.

### Technicalities

Used the following command to cherry-pick changes related to the `IR` from #7580 
```
git format-patch --subject-prefix=ir --stdout master..newrels -- ibis/common ibis/formats ibis/expr ibis/tests/expr ibis/*.py | git am -3 -k
```

### TODOs

These follow-ups must be addressed before doing a release:
- https://github.com/ibis-project/ibis/issues/7820
- https://github.com/ibis-project/ibis/issues/7819

### CODE REVIEW

File: ibis/tests/expr/test_table.py
Comment (kszucs) [inline ibis/tests/expr/test_table.py:1484]: this should be resolved now

File: ibis/expr/types/joins.py
Comment (kszucs) [inline ibis/expr/types/joins.py:153]: Need to expose the full table expr API

File: ibis/expr/operations/relations.py
Comment (kszucs) [inline ibis/expr/operations/relations.py:417]: Remove this.

File: ibis/expr/operations/relations.py
Comment (kszucs) [inline ibis/expr/operations/relations.py:399]: This must be `values` instead of `fields`, doesn't have a test case for it just in the SQL backends, perhaps we could move the entire op there?

File: ibis/expr/analysis.py
Comment (cpcloud) [inline ibis/expr/analysis.py:1]: image

File: ibis/expr/types/joins.py
Comment (cpcloud) [inline ibis/expr/types/joins.py:135]: Before we merge, let's copy the `Table.select` signature and docstring here and do the same for the rest of the APIs.

File: ibis/expr/types/relations.py
Comment (cpcloud) [inline ibis/expr/types/relations.py:102]: If we're covering it by other means, then I think it's fine to leave it as is unless you have specific cases you want to test that have no overlap with hitting this function through the user APIs.

File: ibis/expr/types/relations.py
Comment (cpcloud) [inline ibis/expr/types/relations.py:156]: This needs some elaboration, because it underlies a ton of the changes to generated SQL.

For example:

1. Why is the replace valid?
2. In what cases is it **not** valid?

File: ibis/expr/types/relations.py
Comment (cpcloud) [inline ibis/expr/types/relations.py:171]: Why are we changing the name back to `TableExpr`?

File: ibis/expr/types/relations.py
Comment (cpcloud) [inline ibis/expr/types/relations.py:434]: Is this a new API? If so, it should go into a separate PR.

Try to avoid introducing any new public APIs in this PR, it makes it much harder to review.

Since you're using it below, make it private for now to speed up the review.

File: ibis/expr/types/relations.py
Comment (cpcloud) [inline ibis/expr/types/relations.py:822]: Changing the default is fine, but getting rid of `None` support is a user facing change. What if someone is calling `group_by(by=None, metrics=...)`? That's a legitimate call. Why break it?

File: ibis/expr/types/relations.py
Comment (cpcloud) [inline ibis/expr/types/relations.py:1066]: Still a TODO?

File: ibis/expr/types/relations.py
Comment (cpcloud) [inline ibis/expr/types/relations.py:1072]: What is the actual TODO here?

File: ibis/expr/types/relations.py
Comment (cpcloud) [inline ibis/expr/types/relations.py:1846]:
```suggestion
        return self.select(**node.fields, **values)
```

Just chuck 'em all in there in one call!

File: ibis/expr/types/relations.py
Comment (cpcloud) [inline ibis/expr/types/relations.py:2940]: Ok, but **why**?

File: ibis/expr/types/relations.py
Comment (cpcloud) [inline ibis/expr/types/relations.py:2945]: This comment doesn't seem that helpful, it's describing the code below which isn't obfuscated in any way by indirection.

File: ibis/expr/types/relations.py
Comment (cpcloud) [inline ibis/expr/types/relations.py:2996]: This is also describing **what** rather than **why**.

File: ibis/tests/expr/test_table.py
Comment (cpcloud) [inline ibis/tests/expr/test_table.py:252]: Does this need to be fixed?

File: ibis/tests/expr/test_table.py
Comment (cpcloud) [inline ibis/tests/expr/test_table.py:722]: Does this still need to be revisited?

File: ibis/tests/expr/test_table.py
Comment (cpcloud) [inline ibis/tests/expr/test_table.py:753]: Still a TODO?

File: ibis/tests/expr/test_table.py
Comment (cpcloud) [inline ibis/tests/expr/test_table.py:1193]: Ah, this answers my question above!

**Review by cpcloud — CHANGES_REQUESTED** [review]
Lots of great stuff here!

I think the PR description needs some prose :)

Here's some possible questions to answer in the description:

- How are you solving the chained join problem?
- What are the intended semantics of table references: i.e., what constitutes an integrity error?
- How does the subquery flattening and column pruning work?
- What still needs work and whether it can be done in a follow up
- Are there any expected or known breakages to user code

Ideally most of the core maintainers can read the description and use it as a guide for the PR.

File: ibis/expr/operations/reductions.py
Comment (kszucs) [reply ibis/expr/operations/reductions.py:219]: Ooops, will do!

File: ibis/expr/operations/core.py
Comment (kszucs) [reply ibis/expr/operations/core.py:42]: Yes, I'm going to leave that note and later we can benchmark its effect.

File: ibis/expr/rewrites.py
Comment (kszucs) [reply ibis/expr/rewrites.py:217]: Not yet, removing.

Comment (kszucs) [reply ibis/expr/rewrites.py:217]: Removed.

File: ibis/expr/decompile.py
Comment (kszucs) [reply ibis/expr/decompile.py:87]: Removed.

File: ibis/expr/tests/snapshots/test_sql/test_parse_sql_table_alias/decompiled.py
Comment (kszucs) [reply ibis/expr/tests/snapshots/test_sql/test_parse_sql_table_alias/decompiled.py:9]: Yep, I love it as well.

File: ibis/expr/types/core.py
Comment (kszucs) [reply ibis/expr/types/core.py:355]: Yes, I'm going to move this optimize method to `ibis.expr.rewrites.simplify()` or something like that.

File: ibis/expr/operations/reductions.py
Comment (kszucs) [reply ibis/expr/operations/reductions.py:331]: Restoring.

File: ibis/expr/operations/sortkeys.py
Comment (kszucs) [reply ibis/expr/operations/sortkeys.py:31]: I had no intention to update it in this PR, but I'm leaving the todo note.

File: ibis/expr/rewrites.py
Comment (kszucs) [reply ibis/expr/rewrites.py:83]: You tell me :) 
Are we going to need them in the backends?

Comment (cpcloud) [reply ibis/expr/rewrites.py:83]: Hm 🤔 I'm not sure they'll be needed in this exact form since a base class implementation of `visit_$OP` might suffice.

Let's leave 'em for now :)

File: ibis/expr/types/generic.py
Comment (kszucs) [reply ibis/expr/types/generic.py]: Resolved.

File: ibis/expr/rewrites.py
Comment (kszucs) [reply ibis/expr/rewrites.py:83]: Yeah, these seem to be backend specific enough so they should either be removed or moved to the respective backends. I'm going to delete these, we can easily restore them.

Comment (kszucs) [reply ibis/expr/rewrites.py]: Removed them, we can restore if they are needed.

File: ibis/expr/types/joins.py
Comment (kszucs) [reply ibis/expr/types/joins.py:81]: Was easy enough. I introduced a `JoinKind` type and reused it 3 places.

File: ibis/expr/types/relations.py
Comment (kszucs) [reply ibis/expr/types/relations.py:172]: I prototyped everything in a single file at the beginning and I had name collisions. Restoring it.

Comment (kszucs) [reply ibis/expr/types/relations.py]: Restored.

File: ibis/expr/operations/relations.py
Comment (kszucs) [reply ibis/expr/operations/relations.py:420]: Removed.

File: ibis/expr/operations/relations.py
Comment (kszucs) [reply ibis/expr/operations/relations.py:402]: Updated.

File: ibis/expr/operations/temporal_windows.py
Comment (kszucs) [reply ibis/expr/operations/temporal_windows.py:15]: Removed.

File: ibis/tests/expr/test_table.py
Comment (kszucs) [reply ibis/tests/expr/test_table.py:722]: We don't eagerly rewrite anymore because that would break column dereferencing entirely. So I'm going to remove this test case. We can do the same using the rewrites BTW.

File: ibis/expr/types/generic.py
Comment (cpcloud) [reply ibis/expr/types/generic.py:1198]: This comment is not a blocker btw, so I'm going to resolve it.

File: ibis/expr/types/joins.py
Comment (cpcloud) [reply ibis/expr/types/joins.py:25]: Also not a blocker, resolving.

File: ibis/tests/expr/test_analysis.py
Comment (kszucs) [reply ibis/tests/expr/test_analysis.py:15]: I am not sure what it was testing but I restored the test case and updated according to the new join modeling.

File: ibis/expr/types/core.py
Comment (kszucs) [reply ibis/expr/types/core.py]: Moved, and also removed the flag which was a remnant from the times where I tried to apply eager rewrites.

File: ibis/expr/types/relations.py
Comment (kszucs) [reply ibis/expr/types/relations.py:824]: Ohhh, this is from the pre-experimentation phase. Going to restore the `None` support but keep the new default. BTW `None` is a valid value, it gets coerced to a null literal.

Comment (cpcloud) [reply ibis/expr/types/relations.py:824]: >  BTW None is a valid value, it gets coerced to a null literal.

That's a breaking change I think, because previously it was being used as a sentinel to mean "no grouping keys"

File: ibis/expr/types/relations.py
Comment (cpcloud) [reply ibis/expr/types/relations.py:2926]: What I mean here is just to add a slightly more informative comment about why this is the case.

File: ibis/expr/types/relations.py
Comment (cpcloud) [reply ibis/expr/types/relations.py:2982]: I'll throw out a suggested replacement:

> `by` is an argument that comes from pandas, which for pandas was a convenient way to perform a standard join before the asof join, so we implement there here for consistency across backends.

File: ibis/expr/types/relations.py
Comment (kszucs) [reply ibis/expr/types/relations.py]: Restored, added a test too.

File: ibis/expr/types/relations.py
Comment (kszucs) [reply ibis/expr/types/relations.py:2929]: Added.

File: ibis/tests/expr/test_table.py
Comment (kszucs) [reply ibis/tests/expr/test_table.py:253]: The API is less restrictive now, if we are okay with that then no, otherwise yes.

File: ibis/expr/types/joins.py
Comment (kszucs) [reply ibis/expr/types/joins.py:128]: Ooops, probably lost somewhere down the road. Updating.

File: ibis/tests/expr/test_table.py
Comment (cpcloud) [reply ibis/tests/expr/test_table.py:253]: Seems fine to me!

File: ibis/tests/expr/test_table.py
Comment (kszucs) [reply ibis/tests/expr/test_table.py:1197]: I would prefer to defer it to a follow-up to unblock the backend PRs.

File: ibis/expr/types/joins.py
Comment (kszucs) [reply ibis/expr/types/joins.py:148]: I could use some suggestions here. Actually we should override `__getattribute__` to ensure that `self.finish()` gets called in all methods. Any suggestions for that?

**Review by gforsyth — COMMENTED** [review]
small things.

I think all of our future selves would benefit from writing up the join-chain-deferencing ideas and putting them somewhere in the developer docs.
Maybe under a "Design" or "Internals" section?

Non-blocking, and a lot of what would go in that doc is already present in docstrings here, but I think it would be good to collect it all and then add some more detail.

File: ibis/expr/types/joins.py
Comment (kszucs) [reply ibis/expr/types/joins.py:212]: Done.

File: ibis/expr/types/generic.py
Comment (kszucs) [reply ibis/expr/types/generic.py:16]: Isn't ruff supposed to reorder these?

Comment (kszucs) [reply ibis/expr/types/generic.py:16]: Updated.

File: ibis/expr/operations/relations.py
Comment (kszucs) [reply ibis/expr/operations/relations.py:110]: Fixed and added test case.

File: ibis/expr/types/relations.py
Comment (cpcloud) [inline ibis/expr/types/relations.py:2983]: How does this occur coming from a user-constructed expression? How can the left side be a join chain and the next call ends up here?

File: ibis/tests/expr/test_selectors.py
Comment (kszucs) [reply ibis/tests/expr/test_selectors.py]: Added support for it in bind()

File: ibis/expr/types/relations.py
Comment (cpcloud) [inline ibis/expr/types/relations.py:167]: What's an example of this using the user API?

File: ibis/expr/tests/test_newrels.py
Comment (cpcloud) [inline ibis/expr/tests/test_newrels.py:1054]: Since there's a lot of the same letter floating around, it would really help readability IMO if you added a couple comments stating what each new `aaaaaa` variable is meant to test.

**Review by cpcloud — COMMENTED** [review]
Lookin' good!

Can you make sure to capture the collision implementation in the PR description? It could even go in a follow up IMO.

File: ibis/expr/types/generic.py
Comment (gforsyth) [reply ibis/expr/types/generic.py]: It can, but `ruff format` doesn't do that, you need to do something like `ruff check . --select I001 --select I002 --fix`

File: ibis/tests/expr/test_table.py
Comment (kszucs) [reply ibis/tests/expr/test_table.py:253]: Created a follow-up issue for this https://github.com/ibis-project/ibis/issues/7819

File: ibis/expr/types/joins.py
Comment (kszucs) [reply ibis/expr/types/joins.py]: Fixed.

File: ibis/expr/tests/test_newrels.py
Comment (kszucs) [reply ibis/expr/tests/test_newrels.py:1054]: These are just API variant of the same call. Anyway, I would rather defer the rewrite of tests to unblock the follow-up work.

**Review by cpcloud — APPROVED** [review]

**Comment by cpcloud** [comment]
🎉 🎉 🚀 🤖

**Pull request merged**
